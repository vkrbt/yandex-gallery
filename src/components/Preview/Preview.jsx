import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Img } from '../Img/Img';

const defaultState = {
  transform: 'translate3d(0, 0, 0)',
  transitionDuration: 0,
};
export class Preview extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      imagePos: null,
      ...defaultState,
    };
    const imagePos = props.images.findIndex(image => props.currentId === image.id);
    const isPrevImageExist = !!this.props.images[imagePos - 1];
    const isNextImageExist = !!this.props.images[imagePos + 1];
    this.state = {
      isPrevImageExist,
      imagePos,
      isNextImageExist,
    };

    this.handlePrevImage = this.handlePrevImage.bind(this);
    this.handleNextImage = this.handleNextImage.bind(this);
    this.handleKeyUp = this.handleKeyUp.bind(this);
    this.handleTouchStart = this.handleTouchStart.bind(this);
    this.handleTouchMove = this.handleTouchMove.bind(this);
    this.handleTouchEnd = this.handleTouchEnd.bind(this);
    this.handleImageNotChanged = this.handleImageNotChanged.bind(this);
    this.createTrackRef = this.createTrackRef.bind(this);
  }

  componentDidMount() {
    document.addEventListener('keydown', this.handleKeyUp);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleKeyUp);
  }

  handleKeyUp(e) {
    if (e.key === 'ArrowLeft') {
      this.handlePrevImage();
    }
    if (e.key === 'ArrowRight') {
      this.handleNextImage();
    }
  }

  handlePrevImage() {
    if (this.props.images[this.state.imagePos - 1]) {
      const isPrevImageExist = !!this.props.images[this.state.imagePos - 2];
      this.setState({
        transitionDuration: this.props.transitionDuration,
        transform: `translate3d(100%, 0, 0)`,
      });
      clearTimeout(this.timeoutId);
      this.timeoutId = setTimeout(() => {
        this.setState({
          isPrevImageExist,
          imagePos: this.state.imagePos - 1,
          isNextImageExist: true,
          ...defaultState,
        });
      }, this.props.transitionDuration);
    } else {
      this.handleImageNotChanged();
    }
  }

  handleNextImage() {
    if (this.state.imagePos + 2 === this.props.images.length) {
      this.props.getNext();
    }
    if (this.props.images[this.state.imagePos + 1]) {
      const isNextImageExist = !!this.props.images[this.state.imagePos + 2];
      this.setState({
        transitionDuration: this.props.transitionDuration,
        transform: `translate3d(-100%, 0, 0)`,
      });
      clearTimeout(this.timeoutId);
      this.timeoutId = setTimeout(() => {
        this.setState({
          isPrevImageExist: true,
          imagePos: this.state.imagePos + 1,
          isNextImageExist,
          ...defaultState,
        });
      }, this.props.transitionDuration);
    } else {
      this.handleImageNotChanged();
    }
  }

  handleImageNotChanged() {
    this.setState({
      transitionDuration: this.props.transitionDuration,
      transform: `translate3d(0, 0, 0)`,
    });
    clearTimeout(this.timeoutId);
    this.timeoutId = setTimeout(() => {
      this.setState({
        transitionDuration: 0,
      });
    }, this.props.transitionDuration);
  }

  handleTouchStart(e) {
    [this.firstTouch] = e.touches;
  }

  handleTouchMove(e) {
    const [currentTouch] = e.touches;
    this.lastTouch = currentTouch;
    if (this.touchIsVertical) {
      e.stopPropagation();
      e.preventDefault();
      const translatePosition = currentTouch.clientX - this.firstTouch.clientX;
      this.setState({
        transform: `translate3d(${translatePosition}px, 0, 0)`,
      });
      return;
    }
    const horizontalDistance = Math.abs(currentTouch.clientX - this.firstTouch.clientX);
    const verticalDistance = Math.abs(currentTouch.clientY - this.firstTouch.clientY);
    if (verticalDistance < horizontalDistance) {
      e.stopPropagation();
      e.preventDefault();
      this.touchIsVertical = true;
    }
  }

  handleTouchEnd() {
    if (this.lastTouch) {
      this.touchIsVertical = false;
      const shouldImageBeChanged = Math.abs(this.firstTouch.clientX - this.lastTouch.clientX) > 100;
      if (shouldImageBeChanged) {
        if (this.firstTouch.clientX < this.lastTouch.clientX) {
          this.handlePrevImage();
        } else {
          this.handleNextImage();
        }
      } else {
        this.handleImageNotChanged();
      }
    }
    this.lastTouch = null;
  }

  createTrackRef(track) {
    this.track = track;
  }

  render() {
    const currentImage = this.state.imagePos !== null ? this.props.images[this.state.imagePos] : null;
    const { isPrevImageExist, isNextImageExist, transform, transitionDuration } = this.state;
    const trackStyle = { transitionDuration: `${transitionDuration}ms`, transform };

    const prevImage = isPrevImageExist ? this.props.images[this.state.imagePos - 1] : null;
    const nextImage = isNextImageExist ? this.props.images[this.state.imagePos + 1] : null;
    return (
      <div className="preview">
        {isPrevImageExist ? (
          <button className="preview__control preview__control_prev" onClick={this.handlePrevImage} />
        ) : null}
        <div
          className="preview__track"
          onTouchStart={this.handleTouchStart}
          onTouchMove={this.handleTouchMove}
          onTouchEnd={this.handleTouchEnd}
          onTouchCancel={this.handleTouchEnd}
          ref={this.createTrackRef}
          style={trackStyle}
        >
          {[
            <div key={prevImage ? prevImage.id : -1} className="preview__image-wrapper preview__image-wrapper_prev">
              {prevImage ? <Img className="preview__image_prev" src={prevImage.urls.regular} /> : null}
            </div>,
            <div key={currentImage.id} className="preview__image-wrapper">
              <Img className="preview__image" src={currentImage.urls.regular} />
            </div>,
            <div key={nextImage ? nextImage.id : 1} className="preview__image-wrapper preview__image-wrapper_next">
              {nextImage ? <Img className="preview__image_next" src={nextImage.urls.regular} /> : null}
            </div>,
          ]}
        </div>
        {isNextImageExist ? (
          <button className="preview__control preview__control_next" onClick={this.handleNextImage} />
        ) : null}
      </div>
    );
  }
}

Preview.defaultProps = {
  transitionDuration: 200,
};

Preview.propTypes = {
  images: PropTypes.arrayOf(
    PropTypes.shape({
      urls: PropTypes.shape({
        regular: PropTypes.string.isRequired,
      }).isRequired,
    }).isRequired,
  ).isRequired,
  currentId: PropTypes.string,
  transitionDuration: PropTypes.number,
  getNext: PropTypes.func.isRequired,
};
