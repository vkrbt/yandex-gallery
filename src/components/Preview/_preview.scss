$control_bg_color: #333;
$control_bg_opacity: 0.7;
.preview {
  position: relative;

  overflow: hidden;

  width: 100%;
  height: 100%;

  outline: none;
  &__track {
    position: relative;

    display: flex;
    align-items: center;
    justify-content: center;

    width: 100%;
    height: 100%;
  }
  &__image {
    max-width: 100%;
    max-height: 100%;
  }
  &__image-wrapper_prev,
  &__image-wrapper_next {
    position: absolute;

    display: flex;
    align-items: center;
    justify-content: center;

    width: 100%;
    height: 100%;
  }
  &__image-wrapper_prev {
    right: 100%;
  }
  &__image-wrapper_next {
    left: 100%;
  }
  &__image_prev,
  &__image_next {
    max-width: 100%;
    max-height: 100%;
  }
  &:hover &__control {
    opacity: 0.5;
  }
  &__control {
    position: absolute;
    z-index: 1;
    top: 0;

    width: 100px;
    height: 100%;
    padding: 0;

    cursor: pointer;
    transition: 0.2s ease-out;

    opacity: 0;
    outline: none;
    @media screen and (max-width: 768px) {
      display: none;
    }
    @media (pointer: coarse) and (min-width: 768px) {
      display: block;

      opacity: 0.5;
    }
    &:hover {
      opacity: 1 !important;
    }
    &::after {
      position: absolute;

      display: block;

      width: 15px;
      height: 15px;

      content: '';

      border: 1px solid #fff;
      border-bottom-color: transparent;
      border-left-color: transparent;
    }
    &_prev {
      left: 0;

      background: linear-gradient(-90deg, rgba($control_bg_color, 0), rgba($control_bg_color, $control_bg_opacity));
      &::after {
        top: 50%;
        left: 50%;

        transform: translate(-30%, -50%) rotate(-135deg);
      }
    }
    &_next {
      right: 0;

      background: linear-gradient(90deg, rgba($control_bg_color, 0), rgba($control_bg_color, $control_bg_opacity));
      &::after {
        top: 50%;
        right: 50%;

        transform: translate(30%, -50%) rotate(45deg);
      }
    }
  }
}
