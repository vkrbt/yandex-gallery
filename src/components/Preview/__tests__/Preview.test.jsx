import React from 'react';
import { shallow } from 'enzyme';
import { Preview } from '../Preview';

const images = Array.from({ length: 5 }, (item, i) => ({
  id: `${i + 1}`,
  urls: {
    regular: `url${i + 1}`,
  },
}));
const props = {
  images,
  currentId: images[0].id,
  getNext: jest.fn(),
};

describe('InfiniteScroll', () => {
  it('should being rendered normally', () => {
    const wrapper = shallow(<Preview {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
