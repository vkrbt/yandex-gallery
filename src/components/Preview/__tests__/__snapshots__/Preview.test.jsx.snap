// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteScroll should being rendered normally 1`] = `
<div
  className="preview"
>
  <div
    className="preview__track"
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "transform": undefined,
        "transitionDuration": "undefinedms",
      }
    }
  >
    <div
      className="preview__image-wrapper preview__image-wrapper_prev"
      key="-1"
    />
    <div
      className="preview__image-wrapper"
      key="1"
    >
      <Img
        alt=""
        className="preview__image"
        src="url"
      />
    </div>
    <div
      className="preview__image-wrapper preview__image-wrapper_next"
      key="2"
    >
      <Img
        alt=""
        className="preview__image_next"
        src="url"
      />
    </div>
  </div>
  <button
    className="preview__control preview__control_next"
    onClick={[Function]}
  />
</div>
`;

exports[`InfiniteScroll should change image by calling handleNextImage 1`] = `
<div
  className="preview"
>
  <div
    className="preview__track"
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "transform": undefined,
        "transitionDuration": "undefinedms",
      }
    }
  >
    <div
      className="preview__image-wrapper preview__image-wrapper_prev"
      key="-1"
    />
    <div
      className="preview__image-wrapper"
      key="1"
    >
      <Img
        alt=""
        className="preview__image"
        src="url"
      />
    </div>
    <div
      className="preview__image-wrapper preview__image-wrapper_next"
      key="2"
    >
      <Img
        alt=""
        className="preview__image_next"
        src="url"
      />
    </div>
  </div>
  <button
    className="preview__control preview__control_next"
    onClick={[Function]}
  />
</div>
`;
